libbeat/parser/parser.go

```diff
@@ -35,20 +35,23 @@ var (
 
 // parser transforms or translates the Content attribute of a Message.
 // They are able to aggregate two or more Messages into a single one.
-type parser interface {
+type Parser interface {
 	io.Closer
 	Next() (reader.Message, error)
 }
 
-type parserConfig struct {
-	maxBytes       int
-	lineTerminator readfile.LineTerminator
+type CommonConfig struct {
+	MaxBytes       int
+	LineTerminator readfile.LineTerminator
 }
 
-func newParsers(in reader.Reader, pCfg parserConfig, c []common.ConfigNamespace) (parser, error) {
-	p := in
+type Config struct {
+	pCfg    CommonConfig
+	parsers []common.ConfigNamespace
+}
```

Comment:
Maybe consider to implement `Unpack` on the config. Then it can be used directly with `go-ucfg` unpack like this (we have a similar pattern for TLS and I just introduced one for http):

```
type myConfig struct {

  Parser parser.Config `config:",inline"`
}
```

with

```
func (c *Config) Unpack(cfg *common.Config) error {
  tmp := struct {
    Common CommonConfig `config:",inline"`
    Parsers []common.ConfigNamespace `config:"parser"`
  }{}

  if err := cfg.Unpack(&tmp); err != nil {
    return err
  }

  newC, err := NewConfig(tmp.Common, tmp.Parsers)
  if err != nil {
    return err
  }

  *c = *newC
  return nil
}
```

That way if we have an error, the full name of the setting that failed to parse will be pointed out.

Plus users can use it like:

```
var settings myConfig
if err := cfg.Unpack(&settings); err != nil {
  return err
}

reader := ...
contentsReader := settings.Parser.Create(reader)
```